pkg/storage/replica.go

```diff
@@ -1995,7 +1995,13 @@ func (r *Replica) updateTimestampCache(
 	}
 
 	tc := r.store.tsCache
+	// Update the timestamp cache using the timestamp at which the batch
+	// was executed. Note this may have moved forward from ba.Timestamp,
+	// as when the request is retried locally on WriteTooOldErrors.
 	ts := ba.Timestamp
+	if br != nil {
+		ts = br.Timestamp
```

Comment:
@spencerkimball or @bdarnell, do you know what makes this safe in relation to lease transfers? Specifically, a batch goes through the following steps:
```
1. update the local node's clock with `ba.Timestamp` when passing through the Store
2. under lock, we check whether a lease transfer is in flight
2a. if so, wait for the lease to stabilize
3. if not, evaluate the batch without ever coordinating with possible lease transfers again
4. bump timestamp cache with batch's timestamp (this code)
```

From the lease transfers point of view:
```
1. lock
2. grab a local clock reading
3. start transfer with that time, which will be used as the new leaseholder's tscache lower bound
4. unlock
```

Since this timestamp is larger than the original `ba.Timestamp`, isn't it possible that it's contribution to the timestamp cache is missed? Are we relying on some implicit assumption that this timestamp will never be bumped past `clock.Now()`? Is that a fair assumption to make in all cases? Are we assuming that this timestamp will always be based on existing data on the Range, which in turn must be at a lower timestamp than the clock on the Store that houses it? If so, I don't see how we're confident about that conclussion, given that we only set transactional batches' timestamps [to the txn's orig timestamp](https://github.com/cockroachdb/cockroach/blob/master/pkg/roachpb/batch.go#L41) (where it reads) and we don't include a timestamp when resolving intents.

Reply:
Yes, I think we're implicitly relying on an assumption that any KV data that might bump the response timestamp will have been through some path that would have updated the HLC. I don't know whether that's always valid. It sounds like we might be missing some cases to pull timestamps out of the request (like in intent resolution).

Reply:
It's safe because the `br.Timestamp` is only ever forwarded to the timestamp of a key written to the current range. In other words, the local clock must already be at or past any value of key written to the range. You might want to augment the comment here.

Reply:
This thought experiment blossomed into the beautiful issue we see in https://github.com/cockroachdb/cockroach/issues/36431. Over there we note that even if we did update the leaseholder's HLC clock on intent resolution, we'd still run into issues with transaction uncertainty when intents are moved forward. 

We'll need to address that issue by enforcing clock updates during txn refreshes. In addition, we should do more to ensure that clock updates are never missed, even when `BatchRequests` don't explicitly set a header timestamp. Trying to address this on a case-by-case basis (e.g. https://github.com/cockroachdb/cockroach/pull/35297 and https://github.com/cockroachdb/cockroach/pull/36307) seems like a fraught effort. There has been some discussion about adding a gRPC interceptor to deal with clock updates at a lower-level than in the `BatchRequest` domain.